### PR TITLE
added a basic test to run the oracle

### PR DESCRIPTION
**Pull request by loocapro**
Added an e2e test that:

1. spawns 2 oracles
2. make them peer
3. expects signed tick data

Depends on: https://github.com/paradigmxyz/reth-exex-examples/pull/33

### CODE REVIEW

File: oracle/src/main.rs [hunk 1]
```diff
@@ -67,3 +98,51 @@ fn main() -> eyre::Result<()> {
...
+
+        // // make them connect
+        // let (peer_1, addr_1) = (network_1.peer_id(), network_1.local_addr());
+        // let (peer_2, addr_2) = (network_2.peer_id(), network_2.local_addr());
+        // network_1.add_peer(*peer_1, addr_1);
+        // network_2.add_peer(*peer_2, addr_2);
+
+        // let mut events_1 = network_1.event_listener();
+
+        // let expected_peer_2 = loop {
```
Comment (mattsse) [inline oracle/src/main.rs:133]: what's missing here, do we need this?

Comment (loocapro) [reply oracle/src/main.rs:133]: > what's missing here, do we need this?

Yup, I am a bit struggling with passing each test `ExExContext` to each oracle instance to finalise this e2e as when you pass it down it gets dropped and the NetworkManager with it.

I think it needs to be restructured in order to be passed around and kept alive as it also has the `ExExNotifications` which requires to be mut.

Do you have any recommendations @mattsse ?

Comment (mattsse) [reply oracle/src/main.rs:133]: where does it get dropped, not really following.

cc @shekhirin testing utils need some work here

**Review by shekhirin — CHANGES_REQUESTED** [review]
Can we merge main into this branch, so that the changes from Reth bump aren't visible anymore?

File: Cargo.toml [hunk 2]
```diff
@@ -39,6 +39,7 @@ reth-revm = { git = "https://github.com/paradigmxyz/reth" }
 reth-transaction-pool = { git = "https://github.com/paradigmxyz/reth" }
 reth-tracing = { git = "https://github.com/paradigmxyz/reth" }
 
+
```
Comment (shekhirin) [inline Cargo.toml:42]:
```suggestion
```

File: Cargo.toml [hunk 3]
```diff
@@ -69,3 +70,6 @@ eyre = "0.6"
 # testing
 reth-exex-test-utils = { git = "https://github.com/paradigmxyz/reth" }
 reth-testing-utils = { git = "https://github.com/paradigmxyz/reth" }
+
+
+
```
Comment (shekhirin) [inline Cargo.toml:75]:
```suggestion
```

File: oracle/Cargo.toml [hunk 4]
```diff
@@ -19,6 +19,7 @@ reth-node-ethereum.workspace = true
 reth-primitives.workspace = true
 reth-tracing.workspace = true
 reth.workspace = true
+reth-tokio-util = { git = "https://github.com/paradigmxyz/reth" }
```
Comment (shekhirin) [inline oracle/Cargo.toml:22]: we need to move this to workspace level

File: oracle/Cargo.toml [hunk 5]
```diff
@@ -34,17 +35,19 @@ enr = "0.12"
 # async
 futures.workspace = true
 tokio-stream.workspace = true
-tokio-tungstenite = "0.23"
+tokio-tungstenite = { version = "0.23", features = ["native-tls"] }
 tokio.workspace = true
+rand = "0.8.5"
+
+thiserror = "1"
```
Comment (shekhirin) [inline oracle/Cargo.toml:42]: can you please put it back under `# misc`?

File: oracle/src/main.rs (hunk 1 above)
Comment (loocapro) [reply oracle/src/main.rs]: Yes it's resolved, with https://github.com/paradigmxyz/reth/pull/11907

thanks @shekhirin

File: oracle/src/main.rs [hunk 6]
```diff
@@ -67,3 +80,108 @@ fn main() -> eyre::Result<()> {
...
+        assert_eq!(expected_peer_1, *peer_1);
+
+        // expect signed data
+        let signed_ticker_1 = broadcast_stream_1.next().await.unwrap().unwrap();
+        assert_eq!(signed_ticker_1.ticker.symbol, "BTCUSDT");
+        assert_eq!(signed_ticker_1.signer, signer_1);
+
+        let signed_ticker_2 = broadcast_stream_2.next().await.unwrap().unwrap();
+        assert_eq!(signed_ticker_2.ticker.symbol, "BTCUSDT");
+        assert_eq!(signed_ticker_2.signer, signer_2);
```
Comment (shekhirin) [inline oracle/src/main.rs:185]: so far this tests that both instances received an item from the stream, but it doesn't test if they actually exchange messages as peers after connecting. Can we do that as well?

**Review by shekhirin — APPROVED** [review]
https://github.com/paradigmxyz/reth-exex-examples/pull/26#discussion_r1821039789 can be addressed in a separate PR, so LGTM as well! Thank you!

**Pull request merged**